File: 2_Add_Two_Numbers.py
# 2019/8/13
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        ans = ListNode(0)
        tmp = ans
        while True:
            if l1:
                tmp.val += l1.val
                l1 = l1.next
            if l2:
                tmp.val += l2.val
                l2 = l2.next
            carry = tmp.val // 10
            tmp.val %= 10
            if not l1 and not l2 and carry == 0:
                break
            tmp.next = ListNode(carry)
            tmp = tmp.next
        return ans
    def printListNode(self, l1: ListNode) -> str:
        if not l1:
            return ""
        ans = str(l1.val)
        while l1.next:
            l1 = l1.next
            ans += " -> " + str(l1.val)
        print(ans)
    def arrayToListNode(self, arr1: list) -> ListNode:
        if not arr1:
            return None
        l1 = ListNode(arr1[0])
        l2 = ListNode(arr1[1])
        l2.next = l1
        ans = ListNode(arr1[2])
        ans.next = l2
        return ans


slt=Solution()
# print(slt.printListNode(slt.arrayToListNode([1,2,3])))
l1 = slt.arrayToListNode([9,9,5])
l2 = slt.arrayToListNode([0,0,5])
slt.printListNode(slt.addTwoNumbers(l1,l2))

File: 2_Add_Two_Numbers.go
// 2020/4/19
package main

import (
	"fmt"
	"reflect"
	"time"
)

func main() {
	start := time.Now()
	s1, s2 := []int{2, 4, 3}, []int{5, 6, 4}
	ans := []int{7, 0, 8}
	fmt.Println(reflect.DeepEqual(listNodeToSlice(addTwoNumbers(sliceToListNode(s1), sliceToListNode(s2))), ans))
	s1, s2 = []int{5, 4, 3}, []int{5, 6, 4}
	ans = []int{0, 1, 8}
	fmt.Println(reflect.DeepEqual(listNodeToSlice(addTwoNumbers(sliceToListNode(s1), sliceToListNode(s2))), ans))
	s1, s2 = []int{5, 4, 3, 3}, []int{5, 6, 4, 2}
	ans = []int{0, 1, 8, 5}
	//fmt.Println(listNodeToSlice(addTwoNumbers(sliceToListNode(s1), sliceToListNode(s2))))
	fmt.Println(reflect.DeepEqual(listNodeToSlice(addTwoNumbers(sliceToListNode(s1), sliceToListNode(s2))), ans))
	s1 = []int{1, 2, 3, 4}
	fmt.Println(s1, reflect.DeepEqual(s1, listNodeToSlice(sliceToListNode(s1))))
	s1 = []int{1, 2, 3, 4, 5}
	fmt.Println(s1, reflect.DeepEqual(s1, listNodeToSlice(sliceToListNode(s1))))
	elapsed := time.Now().Sub(start)
	fmt.Println(elapsed)
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head, carry := &ListNode{}, 0
	cur := head
	for l1 != nil || l2 != nil || carry > 0 {
		if l1 != nil {
			carry += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			carry += l2.Val
			l2 = l2.Next
		}
		cur.Val = carry % 10
		carry /= 10
		if l1 != nil || l2 != nil || carry > 0 {
			cur.Next = &ListNode{}
			cur = cur.Next
		}
	}
	return head
}

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func sliceToListNode(input []int) *ListNode {
	head := &ListNode{Val: input[0]}
	cur := head
	input = input[1:]
	for len(input) > 0 {
		cur.Next = &ListNode{}
		cur = cur.Next
		cur.Val = input[0]
		input = input[1:]
	}
	return head
}

func listNodeToSlice(l *ListNode) []int {
	s := []int{}
	for l != nil {
		s = append(s, l.Val)
		l = l.Next
	}
	return s
}
